### DISCUSSION TITLE
Unit conversions in Weather platform

### DISCUSSION THREAD

**Discussion by gjohansson-ST** [home-assistant/architecture#772] [Entity Models] [answered]
Unit conversions for weather platform has never been properly set and while it appears someone made a PR to add them in the platform it has not been finally decided and thereby PR's to address the issues for the respective integrations are not merged in to fix the issues at hand.

Using wind speed as example as it's quite apparent (metric system)
Developer docs says it should be coded as km/h: https://developers.home-assistant.io/docs/core/entity/weather#properties
Core unit system expects m/s: https://github.com/home-assistant/core/blob/dev/homeassistant/util/unit_system.py#L194
Frontend displays it as m/s

What currently happens is that the integration provides wind speed in km/h and there is no conversion so it is displayed falsely in frontend as m/s.

Suggestion
Fix the platform to correctly use the unit system already in place and thereby make the correct conversions to the respective base unit https://github.com/home-assistant/core/blob/dev/homeassistant/util/unit_system.py#L190-L210
Adjust develop docs to match up with unit system https://developers.home-assistant.io/docs/core/entity/weather/#properties

**Comment by rianadon** [comment]
> Unit conversions for weather platform has never been properly set and while it appears someone made a PR to add them in the platform it has not been finally decided and thereby PR's to address the issues for the respective integrations are not merged in to fix the issues at hand.

That's me 😁 

I think the best solution would be to use a combination of the unit system and the entity registry to allow full customization of the units used for weather platforms. I started a second discussion in https://github.com/home-assistant/architecture/discussions/773 with technical details of how we may do that.

And good point, the dev docs need updating.

**Comment by MatthewFlamm** [comment]
Thank you for starting this discussion. The current state of weather entity units is a mess with no clear direction to fix it.

> Fix the platform to correctly use the unit system already in place and thereby make the correct conversions to the respective base unit https://github.com/home-assistant/core/blob/dev/homeassistant/util/unit_system.py#L190-L210

I would prefer that we start enforcing the final units to the existing entity model units, e.g. hPa and inHg for pressure.  The `IMPERIAL_SYSTEM` has units of `psi` for pressure which would be common for pressure in piping, etc., but almost never used for atmospheric pressure, which is typically in `inHg`.  We can create unit systems (one for metric and other for imperial) specifically for weather instead of using the config unit system as is currently implemented.  This would also require reverting the changes to the frontend that also violated the entity model units.

**Comment by emontnemery** [comment]
@rianadon I think the problem with the previous approach was that changes were unexpectedly forced on users.

@gjohansson-ST converting to units specified by the unit system does not really work because many countries use traditional units for wind speed and barometric pressure. For example, users may want wind speed in knots or km/h instead of m/s, and barometric pressure in mmHg instead of hPa.

I therefore suggest that the weather integration, at least initially, takes a very light handed approach where we allow unit configuration for weather entities from frontend in the same way as we now do for sensor entities, but do not enforce automatic conversion.

If you want, we can chat on discord, I'm emontnemery#6618 there

**Reply by gjohansson-ST** [reply]
Would be happy to discuss best options. I'll reach out on discord in short time.

**Reply by rianadon** [reply]
@emontnemery I could be wrong, but by "do not enforce automatic conversion," are you suggesting let each weather integration specify the set of units it wishes to use, before the user might change them?

If so, I think this may be a great idea to sidestep the issue of the unit system for the time being. Using the existing unit system for weather units was a terrible decision in my part and led to the mess of inconsistencies in units we have now.

**Comment by emontnemery** [comment]
> What currently happens is that the integration provides wind speed in km/h and there is no conversion so it is displayed falsely in frontend as m/s.

Is there a frontend issue open for that?

**Reply by gjohansson-ST** [reply]
No idea but for me there is (at least for wind speed) not an issue as it's in m/s which I would expect. 
I say the issue is completely in core and not in frontend but didn't check too much in any other uom's so I'm not 100%

**Comment by emontnemery** [comment, marked as answer]
Proposal:
- Allow users to override the units documented here: https://developers.home-assistant.io/docs/core/entity/weather in a similar way as users can override the units for temperature and pressure sensors (PR https://github.com/home-assistant/core/pull/64366 + https://github.com/home-assistant/frontend/pull/12143)
- Add new state attributes `precipitation_unit_of_measurement`, `pressure_unit_of_measurement`, `temperature_unit_of_measurement`, `visibility_unit_of_measurement`, `wind_speed_unit_of_measurement` to weather states, with allowed units for each new attribute as outlined below.
  - Allowed units for `precipitation_unit_of_measurement`: `mm`, `in`
  - Allowed units for `pressure_unit_of_measurement`: `hPa`, `mbar`, `mmHg`, `inHg`
  - Allowed units for `temperature_unit_of_measurement`: `°C`, `°F`
  - Allowed units for `visibility_unit_of_measurement`: `km`, `mi`
  - Allowed units for `wind_speed_unit_of_measurement`: `km/h`, `mph`, `m/s`

We could also consider renaming `WeatherEntity` properties `precipitation_unit` etc. to `native_..._unit` to make it clear it's the native unit, before unit conversion, following the same pattern as `sensor` and (soon, pending approval of https://github.com/home-assistant/core/pull/73233) `number`, but that's more of a code consistency matter.

**Comment by Jc2k** [comment]
Should we decompose weather into sensor entities? It was brought up when we did that for AQI. My understanding is that weather would not look like it does now if it was added today. Today it probably would just be a bunch of sensors linked together by a device registry entry (with a type of service).

Obviously it's quite a large breaking change, but that aside, would it be a good idea? If it was, we could probably have a transition period.

Personally, i'd find it very useful if some of the state attributes in weather were just sensors. It would be easy to break out the fields I was interested in in lovelace, as graphs and gauges would be more readily available. They'd get device automations. And then they would be able to leverage the existing unit system. Things like https://github.com/home-assistant/core/pull/64366 would end up being supported for free, rather than having to reimplement them.

Other than the breaking change, are there reasons to keep weather using the "old way" of building HA integrations?

**Reply by emontnemery** [reply]
@Jc2k I think it's tricky to break weather up in individual sensors if we want consumers, for example frontend, to be able to present weather data including forecast in a consistent manner. We would then need to define a sensor scheme for a "weather device", including which sensors represent forecasts. `air_quality` is a bit different, especially considering it doesn't support forecasts.

**Reply by Jc2k** [reply]
Maybe that's worth doing. The developer documentation does cover how *not* to represent forecasts (i.e. a `measurement` is not a forecast), but doesn't cover how you should represent a forecast.

Or maybe `weather` covers forecasts, but the "right now" sensors are broken out as sensors.

**Reply by emontnemery** [reply]
I also think it's worth looking into further, but it will require breaking changes and new concepts where a device has a type "weather" with some sort of scheme for which entities may be included in that device and how they related to each other.
That can be explored further, but the small non-breaking change where the units are clarified and made customizable by the user as suggested in https://github.com/home-assistant/architecture/discussions/772#discussioncomment-2936386 can be done first.
